File: src/transfers/PlayerValuation.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Transfers;

/// <summary>
/// Calculates market value for players based on overall rating, age, and club reputation.
/// </summary>
public static class PlayerValuation
{
    /// <summary>
    /// Calculates a player's market value in the transfer market.
    /// </summary>
    public static decimal Calculate(Player player, int clubReputation = 50)
    {
        float overall = player.PrimaryPosition == Position.GK
            ? player.Attributes.GoalkeeperOverall
            : player.Attributes.OutfieldOverall;

        // Exponential base value from overall (65 ovr ≈ 50k, 75 ≈ 150k, 85 ≈ 500k)
        decimal baseValue = (decimal)Math.Pow(overall / 10.0, 3.5) * 15m;

        // Age factor: peak 25-28, young premium, old discount
        decimal ageFactor = GetAgeFactor(player.Age);

        // Club reputation adds 0-30% markup
        decimal repBonus = 1m + clubReputation / 500m;

        decimal value = baseValue * ageFactor * repBonus;

        // Minimum value, rounded to nearest 1000
        return Math.Max(5_000m, Math.Round(value / 1000m) * 1000m);
    }

    /// <summary>
    /// Gets a salary estimate for a player (weekly cost).
    /// </summary>
    public static decimal EstimateWeeklySalary(Player player)
    {
        float overall = player.PrimaryPosition == Position.GK
            ? player.Attributes.GoalkeeperOverall
            : player.Attributes.OutfieldOverall;

        return (decimal)overall * 10m;
    }

    private static decimal GetAgeFactor(int age) => age switch
    {
        <= 17 => 0.6m,
        18 => 0.8m,
        19 => 0.9m,
        20 => 1.0m,
        21 => 1.05m,
        22 => 1.1m,
        23 => 1.15m,
        24 => 1.2m,
        >= 25 and <= 28 => 1.2m,
        29 => 1.1m,
        30 => 0.95m,
        31 => 0.8m,
        32 => 0.65m,
        33 => 0.5m,
        34 => 0.35m,
        _ => 0.25m
    };
}

File: engine/transfers/player-valuation.ts
import { Position } from "@/engine/enums";
import type { Player } from "@/engine/models";
import { goalkeeperOverall, outfieldOverall } from "@/engine/models";

/**
 * Calculates market value for players based on overall rating, age, and club reputation.
 */

/**
 * Calculates a player's market value in the transfer market.
 */
export function calculate(player: Player, clubReputation: number = 50): number {
  const overall =
    player.primaryPosition === Position.GK
      ? goalkeeperOverall(player.attributes)
      : outfieldOverall(player.attributes);

  // Exponential base value from overall (55 ovr ~ 30k, 65 ~ 80k, 75 ~ 300k, 85 ~ 1.2M)
  const baseValue = Math.pow(overall / 10.0, 4.0) * 30;

  // Age factor: peak 25-28, young premium, old discount
  const ageFactor = getAgeFactor(player.age);

  // Club reputation adds 0-30% markup
  const repBonus = 1 + clubReputation / 500;

  const value = baseValue * ageFactor * repBonus;

  // Minimum value, rounded to nearest 1000
  return Math.max(5_000, Math.round(value / 1000) * 1000);
}

/**
 * Gets a salary estimate for a player (weekly cost).
 */
export function estimateWeeklySalary(player: Player): number {
  const overall =
    player.primaryPosition === Position.GK
      ? goalkeeperOverall(player.attributes)
      : outfieldOverall(player.attributes);

  return overall * 10;
}

function getAgeFactor(age: number): number {
  if (age <= 17) return 0.6;
  switch (age) {
    case 18:
      return 0.8;
    case 19:
      return 0.9;
    case 20:
      return 1.0;
    case 21:
      return 1.05;
    case 22:
      return 1.1;
    case 23:
      return 1.15;
    case 24:
      return 1.2;
    case 25:
    case 26:
    case 27:
    case 28:
      return 1.2;
    case 29:
      return 1.1;
    case 30:
      return 0.95;
    case 31:
      return 0.8;
    case 32:
      return 0.65;
    case 33:
      return 0.5;
    case 34:
      return 0.35;
    default:
      return 0.25;
  }
}
